sql/src/main/java/io/crate/planner/operators/Filter.java

```diff
@@ -71,11 +71,8 @@ public static LogicalPlan create(LogicalPlan source, Symbol query) {
         assert query.valueType().equals(DataTypes.BOOLEAN)
             : "query must have a boolean result type, got: " + query.valueType();
 
-        if (query instanceof Literal) {
-            Boolean value = (Boolean) ((Literal) query).value();
-            if (value != null && value) {
-                return source;
-            }
+        if (query.symbolType().isValueSymbol() && ((Literal) query).value() == Boolean.TRUE) {
```

Comment:
I know from the PR title that you're doing this on purpose, but I must ask why :) . This line has a potential NPE and ClassCastException, so is this really better?

Reply:
The `query` cannot be null as we return Literal.BOOLEAN_TRUE in that case (see: https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/expression/operator/AndOperator.java#L133)

The instanceOf is always a non-preferred way, that's why we have the `isValueSymbol()` that shows us if the Symbol is Literal.

Reply:
well..this is a public method, I don't think you can/should assume that you know where all calls will come from (and not pass a null value in) - I'd recommend having a precondition check still (lombok has great annotations for such preconditions.)
And regarding isValueSymbol() telling you that it's of type Literal, I'd say that's not so readable (perhaps if it was isValueLiteral...)

Reply:
The argument in not annotated as @Nullable

Reply:
I think we can assume that the query here won't be null as this under our control ;) (its not a public library api). 
also isValueSymbol() is our preferred way of checking that this is a literal as values are always literals. renaming this to isValueLiteral() doesn't make sense to me.. if this method would indeed be able of taking NULL arguments the preferred way would be to add a @Nullable annotation and then take care of null values. a precondition check doesn't make sense to me here inside internal api's as well as this would result in an error if the value isn't null which should then never happen and should be protected by an assertion instead

Reply:
I think if `isValueSymbol` is immediately followed by a cast to `Literal` a instanceof check would actually be better. It looks more sensible even without knowing how `isValueSymbol` is implemented and is also more resilient towards changes - in case we ever add a type of symbol where `isValueSymbol` is also true, but it's not a literal.
(also, `instance` checks always imply ´non-null` so you'd get the `not null` checks for free)